libsolidity/analysis/TypeChecker.cpp

```diff
@@ -773,35 +773,56 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
 			// Infer type from value.
 			solAssert(!var.typeName(), "");
 			if (
-				valueComponentType->category() == Type::Category::IntegerConstant &&
-				!dynamic_pointer_cast<IntegerConstantType const>(valueComponentType)->integerType()
+				valueComponentType->category() == Type::Category::RationalNumber &&
+				!dynamic_pointer_cast<RationalNumberType const>(valueComponentType)->integerType() &&
```

Comment:
The call to `integerType` triggers an assertion if called on a non-integer. I guess the best thing here is to check that `mobileType()` does not return a null shared pointer.

This could be done in a generic way (not only for RationalNumber category) and then tune the error message in the case we have a RationalNumber.